File: js/__tests__/BST.test.js
// BST测试
const BST = require('../data_structures/BST')

function createTree() {
    const bst = new BST();
    bst.insert(13)
    bst.insert(55)
    bst.insert(12)
    bst.insert(2)
    bst.insert(7)
    bst.insert(19)
    /**
     *       13
     *   12       55
     * 2        19
     *   7     
     */
    return bst
}

describe('BST', () => {
    it('插入 method', () => {
        const bst = new BST();
        bst.insert(10)
        expect(bst.root.data).toEqual(10)
        bst.insert(20)
        bst.insert(3)
        expect(bst.root.right.data).toEqual(20)
        expect(bst.root.left.data).toEqual(3)
        // 第二次插入
        bst.insert(13)
        bst.insert(59)
        bst.insert(4)
        /**
         *         10
         *     3        20
         *       4   13    59
         */
        const leftNode = bst.root.left;
        const rightNode = bst.root.right;
        expect(leftNode.right.show()).toEqual(4)
        expect(rightNode.left.show()).toEqual(13)
        expect(rightNode.right.data).toEqual(59)
    })
    it('中序遍历 method', () => {
        const bst = createTree()
        bst.midOrder(bst.root)
        expect(bst.show()).toEqual([2, 7, 12, 13, 19, 55])
    })
    /**
     *       13
     *   12       55
     * 2        19
     *   7     
     */
    it('先序遍历', () => {
        const bst = createTree();
        bst.preOrder(bst.root);
        expect(bst.show()).toEqual([13, 12, 2, 7, 55, 19]);
    })
    it('后序遍历', () => {
        const bst = createTree();
        bst.postOrder(bst.root)
        expect(bst.show()).toEqual([7, 2, 12, 19, 55, 13])
    })
    it('findMin', () => {
        const bst = createTree();
        expect(bst.findMin()).toEqual(2)
    })
    it('findMax', () => {
        const bst = createTree()
        expect(bst.findMax()).toEqual(55)
    })
    it('find', () => {
        const bst = createTree();
        expect(bst.find(32)).toBeFalsy();
        expect(bst.find(13)).toBeTruthy();
        expect(bst.find(19)).toBeTruthy();
    })
    /**
     *       13
     *   12       55
     * 2        19
     *   7     
     */
    it('remove, 没有找到对应元素', () => {
        const bst = createTree();
        bst.remove(33)
        expect(bst.find(33)).toBeFalsy()
    })
    it('remove, 左侧单叶子节点', () => {
        const bst = createTree();
        expect(bst.find(12)).toBeTruthy()
        bst.remove(12)
        expect(bst.find(12)).toBeFalsy()
        expect(bst.root.left.data).toEqual(2)
        bst.midOrder(bst.root)
        /**
         *       13
         *   2       55
         *     7   19    
         */
        expect(bst.show()).toEqual([2, 7, 13, 19, 55])
    })
    /**
     *       13
     *   12       55
     * 2        19
     *   7 
     *     8   
     */
    it('remove, 右侧子节点', () => {
        const bst = createTree()
        bst.insert(8)
        bst.remove(7)
        expect(bst.root.left.left.right.data).toEqual(8)
    })
    /**
     *       13
     *   12       55
     * 2        19
     *   7     
     */
    it('remove, 双叶子节点', () => {
        const bst = createTree();
        bst.insert(12.5)
        /**
         *          13
         *    12                55
         * 2     12.5        19
         *   7  
         */
        bst.remove(12)
        expect(bst.root.left.data).toEqual(12.5)
        bst.midOrder(bst.root)
        expect(bst.show()).toEqual([2, 7, 12.5, 13, 19, 55])
    })
    it('remove, 嵌套子节点', () => {
        const bst = createTree();
        bst.insert(28)
        bst.insert(27)
        bst.insert(18.7)
        /**
         *       13
         *   12                    55
         * 2             19
         *   7     18.7       28
         *                27
         */
        bst.remove(19)
        bst.midOrder(bst.root)
        expect(bst.show()).toEqual([2, 7, 12, 13, 18.7, 27, 28, 55])
    })
})

File: py/__tests__/BST_test.py
from data_structures.BST import BST
import unittest


# 二叉查找树测试
def _create_tree():
    tree = BST()
    tree.insert(22)
    tree.insert(33)
    tree.insert(14)
    tree.insert(65)
    tree.insert(18)
    tree.insert(6)
    #          22
    #    14         33
    # 6     18          65
    return tree


class TestBST(unittest.TestCase):

    def test_insert(self):
        tree = _create_tree()
        self.assertEqual(tree.root.data, 22)
        self.assertEqual(tree.root.left.data, 14)
        self.assertEqual(tree.root.right.right.data, 65)

    # 中序 左 -> 根 -> 右
    def test_mid_order(self):
        tree = _create_tree()
        tree.mid_order(tree.root)
        self.assertEqual(tree.order_list, [6, 14, 18, 22, 33, 65])

    # 先序 根 -> 左 -> 右
    def test_pre_order(self):
        tree = _create_tree()
        tree.pre_order(tree.root)
        self.assertEqual(tree.order_list, [22, 14, 6, 18, 33, 65])

    # 后序 左 -> 右 -> 根
    def test_post_order(self):
        #          22
        #    14         33
        # 6     18          65
        tree = _create_tree()
        tree.post_order(tree.root)
        self.assertEqual(tree.order_list, [6, 18, 14, 65, 33, 22])

    def test_find_min(self):
        tree = _create_tree()
        self.assertEqual(tree.find_min(), 6)
        tree_only = BST()
        # 考虑临界值情况
        tree_only.insert(7)
        self.assertEqual(tree_only.find_min(), 7)

    def test_find_max(self):
        tree = _create_tree()
        self.assertEqual(tree.find_max(), 65)
